wgsl/index.bs

```diff
@@ -353,59 +353,102 @@ Note: Only a function declaration can contain other declarations.
 
-### Explanation for those familiar with formal type checking ### {#type-check-preamble}
+The values that may result from evaluating a particular expression will always belong to a specific [SHORTNAME] type,
+known as the <dfn noexport>static type</dfn> of the expression.
+The rules of [SHORTNAME] are designed so the static type of an expression depends only on the expression's static context.
 
-Much of it can be bottom-up, like usual.
+A <dfn noexport>type assertion</dfn> is a mapping from some [SHORTNAME] source expression to a [SHORTNAME] type.
+The notation
 
-The interesting bit is that the type of a pointer expression is either straightforward pointer
-type itself, or the pointee type, depending on its [[#pointer-evaluation]] context:
+> *e* : *T*
 
-* In Indexing, Assigning (LValue), and Copying contexts, the pointer expression denotes a pointer value.
-* In a Parameter context:
- * If the parameter type matches the pointer expression's straightforward pointer type, then the expression denotes that pointer type.
- * Otherwise the pointer expression denotes a value of the pointee type, being the value loaded (at that time) from the referenced storage.
-* In a Reading (RValue) context, the pointer expression denotes a value of the pointee type.
+is a type assertion meaning *T* is the static type of [SHORTNAME] expression *e*.
 
-### How to read type-checking rules ### {#type-check-how-to-read}
+<dfn noexport>Type checking</dfn> is the process of mapping each expression
+in a succesfully parsed [SHORTNAME] program to its static type.
 
-A *type assertion* is a mapping from some [SHORTNAME] source expression to an [SHORTNAME] type.
-When this specification has
+Type checking can be performed by recursively applying type rules.
+A <dfn noexport>type rule</dfn> has two parts:
+* A conclusion, stated as a type assertion for an expression.
+    The expression in the type assertion is specified schematically,
+    using *italicized* names to denote sub-expressions
+    or other syntactically-determined parameters.
+* Preconditions, consisting of:
+    * Type assertions for sub-expressions, when there are sub-expressions.
+    * Conditions on the other schematic parameters, if any.
+    * Optionally, other static context.
 
-> *e* : *T*
+A <dfn noexport>type rule applies to an expression</dfn> when:
+* the rule's conclusion matches a valid parse of the expression, and
+* the rule's preconditions are satisfied.
+
+TODO: write an example such as `1+2`, or `3 - a`, where `a` is in-scope of a const declaration with `i32` type.
+
+The type rules are designed so if parsing succeeds, at most one type rule will apply to each expression.
+If a type rule applies to an expression, then the conclusion is asserted, and therefore determines the static type of the expression.
+
+If the static type can be determined for each expression in a [SHORTNAME] source program by applying
+the type rules, then we say the program is <dfn noexport>well-typed</dfn>.
```

Comment:
This seems necessary, but not sufficient. Given that assignments are not expressions for example. Is intended that the following is well-typed?
```
var x : i32 = 1u;
```

Reply:
Right.  Good point.   I've dropped any notion of type-checking full statements.
That probably should be fixed.  E.g.  the test on a for loop or an if statement must be of type bool.

I didn't want to wade into the waters of arguing that each _statement_ must have a type.  But perhaps there's a middle ground to ensure the "type rule" concept also covers statements.

Reply:
Statements do not have "conclusion" types, but they use expressions, which are typed. There is no need to say that "statement must have a type", that would be opposite from what we have (statements don't have types).

Reply:
I agree you don't need statements to have a "result" type, but it seems odd to say the program is well-typed for the example I gave. I think we could add to the definition of well-typed to say expressions and statement type rules and then define the per statement type rules (which we need to do anyways) in the those sections of the spec.

For example, for assignment statements, the LHS and RHS must have the same type. For for loops, the condition expression must be a bool type, etc.

Reply:
yes, precisely, we are on the same page